exp/mutators/azureasomanagedcontrolplane.go

```diff
@@ -0,0 +1,167 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutators
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	asocontainerservicev1 "github.com/Azure/azure-service-operator/v2/api/containerservice/v1api20231001"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	infrav1exp "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha1"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	// ErrNoManagedClusterDefined describes an AzureASOManagedControlPlane without a ManagedCluster.
+	ErrNoManagedClusterDefined = fmt.Errorf("no %s ManagedCluster defined in AzureASOManagedControlPlane spec.resources", asocontainerservicev1.GroupVersion.Group)
+)
+
+// SetManagedClusterDefaults propagates values defined by Cluster API to an ASO ManagedCluster.
+func SetManagedClusterDefaults(asoManagedControlPlane *infrav1exp.AzureASOManagedControlPlane, cluster *clusterv1.Cluster) ResourcesMutator {
+	return func(ctx context.Context, us []*unstructured.Unstructured) error {
+		ctx, _, done := tele.StartSpanWithLogger(ctx, "mutators.SetManagedClusterDefaults")
+		defer done()
+
+		var managedCluster *unstructured.Unstructured
+		var managedClusterPath string
+		for i, u := range us {
+			if u.GroupVersionKind().Group == asocontainerservicev1.GroupVersion.Group &&
+				u.GroupVersionKind().Kind == "ManagedCluster" {
+				managedCluster = u
+				managedClusterPath = fmt.Sprintf("spec.resources[%d]", i)
+				break
+			}
+		}
+		if managedCluster == nil {
+			return reconcile.TerminalError(ErrNoManagedClusterDefined)
+		}
+
+		if err := setManagedClusterKubernetesVersion(ctx, asoManagedControlPlane, managedClusterPath, managedCluster); err != nil {
+			return err
+		}
+
+		if err := setManagedClusterServiceCIDR(ctx, cluster, managedClusterPath, managedCluster); err != nil {
+			return err
+		}
+
+		if err := setManagedClusterPodCIDR(ctx, cluster, managedClusterPath, managedCluster); err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+func setManagedClusterKubernetesVersion(ctx context.Context, asoManagedControlPlane *infrav1exp.AzureASOManagedControlPlane, managedClusterPath string, managedCluster *unstructured.Unstructured) error {
+	_, log, done := tele.StartSpanWithLogger(ctx, "mutators.setManagedClusterKubernetesVersion")
+	defer done()
+
+	capzK8sVersion := strings.TrimPrefix(asoManagedControlPlane.Spec.Version, "v")
+	if capzK8sVersion == "" {
+		// If there is no CAPI opinion, users can do whatever they want.
```

Comment:
I'm a bit confused by this. This statement checks if there is no capz k8s version, but the comment talks about no CAPI opinion. Doesn't the user specify `asoManagedControlPlane.Spec.Version`?

Reply:
I'm lumping CAPI and CAPZ in the same "CAPI" bucket here since `spec.version` is a CAPI contract field. I'll update this comment to clarify that what I really mean here is "any value for version in the embedded ASO resource may be specified."